libraries/chain/include/graphene/chain/types.hpp

```diff
@@ -0,0 +1,106 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#pragma once
+
+#include <graphene/protocol/types.hpp>
+
+namespace graphene { namespace chain {
+
+using namespace graphene::protocol;
+
+enum impl_object_type {
+    impl_global_property_object_type,
+    impl_dynamic_global_property_object_type,
+    impl_reserved0_object_type,      // formerly index_meta_object_type, TODO: delete me
+    impl_asset_dynamic_data_type,
+    impl_asset_bitasset_data_type,
+    impl_account_balance_object_type,
+    impl_account_statistics_object_type,
+    impl_transaction_object_type,
+    impl_block_summary_object_type,
+    impl_account_transaction_history_object_type,
+    impl_blinded_balance_object_type,
+    impl_chain_property_object_type,
+    impl_witness_schedule_object_type,
+    impl_budget_record_object_type,
+    impl_special_authority_object_type,
+    impl_buyback_object_type,
+    impl_fba_accumulator_object_type,
+    impl_collateral_bid_object_type
+};
+
+using global_property_id_type = object_id<implementation_ids, impl_global_property_object_type>;
+using dynamic_global_property_id_type = object_id<implementation_ids, impl_dynamic_global_property_object_type>;
+using asset_dynamic_data_id_type = object_id<implementation_ids, impl_asset_dynamic_data_type>;
+using asset_bitasset_data_id_type = object_id<implementation_ids, impl_asset_bitasset_data_type>;
+using account_balance_id_type = object_id<implementation_ids, impl_account_balance_object_type>;
+using account_statistics_id_type = object_id<implementation_ids, impl_account_statistics_object_type>;
+using transaction_obj_id_type = object_id<implementation_ids, impl_transaction_object_type>;
+using block_summary_id_type = object_id<implementation_ids, impl_block_summary_object_type>;
+using account_transaction_history_id_type = object_id<implementation_ids, impl_account_transaction_history_object_type>;
+using chain_property_id_type = object_id<implementation_ids, impl_chain_property_object_type>;
+using witness_schedule_id_type = object_id<implementation_ids, impl_witness_schedule_object_type>;
+using budget_record_id_type = object_id<implementation_ids, impl_budget_record_object_type>;
+using blinded_balance_id_type = object_id<implementation_ids, impl_blinded_balance_object_type>;
+using special_authority_id_type = object_id<implementation_ids, impl_special_authority_object_type>;
+using buyback_id_type = object_id<implementation_ids, impl_buyback_object_type>;
+using fba_accumulator_id_type = object_id<implementation_ids, impl_fba_accumulator_object_type>;
+using collateral_bid_id_type = object_id<implementation_ids, impl_collateral_bid_object_type>;
```

Comment:
It's a holy war... Dan did it inconsistently, I never do it. When I was a wee little lad in programming school, the style guide textbook said not to do it because it makes editing a massive pain, and I've always agreed. Most style guides I've seen recommend against it, but don't forbid it.

I don't really care, but I do dearly hate maintaining those things.

Reply:
A macro would clean it up a bit, though. I'll play around with it.